### DISCUSSION TITLE
PyTorch Workflow: How does the loss function know what is going on with the model parameters?

### DISCUSSION THREAD

**Discussion by KevinGregory** [mrdbourke/pytorch-deep-learning#714] [Q&A]
I'm trying to understand the training loop. Everything makes sense to me except for the loss function. I understand that we calculate the loss in a given epoch with `loss = loss_fn(y_pred, y_train)`. I'm confused about two things: 
1) how does `loss.backward()` know what the model parameters are? We don't connect it to the model. We define `loss_fn = nn.L1Loss()` and then we put in two tensors into the loss function with `loss = loss_fn(y_pred, y_train)`. Does it know the parameters from the y_pred, y_train? I'm used to scikit learn where these are just vectors, but maybe the tensors have more information attached to them? 

2) How does the optimizer know what the loss is? We never pass the loss function into the optimizer.

**Comment by jagan-lakshmipathy** [comment]
There is no easy answer to your question. Torch builds the computational graph for each computation you perform in the forward pass of the model. Loss function takes in the output variable from the forward pass that references the computational graph. Torch moves backwards (loss.backwards()) from the loss to each gradient in the computational graph using chain rule. Optimizer advances the weights with the gradients at the optimizer.step().

**Comment by shisirkha** [comment]
The `loss.backward()` function in PyTorch uses the computational graph that is created when the forward pass is executed to compute the gradients of the loss with respect to the model parameters. The computational graph is a directed acyclic graph that represents the sequence of operations that are performed during the forward pass. Each node in the graph represents a tensor, and each edge represents a function that transforms the input tensor into the output tensor.

When you call `loss.backward()`, PyTorch uses the chain rule of calculus to compute the gradients of the loss with respect to each tensor in the graph. The gradients are then stored in the `.grad` attribute of each tensor. When you call `optimizer.step()`, the optimizer uses these gradients to update the model parameters.

The optimizer does not need to know the loss function explicitly. Instead, it uses the gradients that are computed by the `loss.backward()` function to update the model parameters. The loss function is only used to compute the loss value, which is then used to compute the gradients.


I hope this explanation helps

**Comment by anurag2506** [comment]
To define what the loss function does and how it affects the model parameters, its essential to understand what loss and model parameters are. Loss is the final resultant tensor that is obtained after making 'n' number of operations on the initial tensor. This initial tensors are the model parameters, ie the weights and biases of every neuron

Lets take an example of a basic MLP:

```
model = torch.nn.Sequential(
    nn.Linear(512,128),
    nn.Linear(128,32),
    nn.Linear(32,8),
    nn.Linear(8,2)
).to(device)
```
The MLP is a series of nested transformations made to the model parameters. 
Here, the model parameters are the perceptrons having weights and biases. When `requires_grad = True` is set, then you have a directed acyclic graph formed that looks like this: 

<img width="803" height="712" alt="image" src="https://github.com/user-attachments/assets/02ca343f-d9c5-4253-8c30-d22581496f0c" />

- When the forward method is called by `y_pred = model(X) `, the weights, biases and activations propagate through the network. 
- Then the loss is calculated between the predicted and true values using any loss function like `loss = nn.MSE() or nn.BCE()` 
- After performing the forward and loss methods, we call the `loss.backward()` to calculate the gradients of the initial tensors throughout the entire network. PyTorch uses the chain rule of calculus to compute the gradients. 
- By calculating these gradients, we know the influence of each model parameter on the loss and how each of them have to be updated accordingly to reduce this loss.
- And we can get these gradients by calling the `.grad` on each of the parameters. 

Hope this helps, cheers!
